### PR TITLE
Boosting search results on download counts.

### PR DESCRIPTION
**Pull request by mhilton**
Boost search results based on download counts.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/634/

**Comment by mhilton** [comment]
This one will be quite tricky to QA. When you have a full system running and have ingested the legacy download counts. Then if you run a search when all other things are equal the higher download numbers should be earlier.

**Comment by urosj** [comment]
:+1: on code.

Did QA:
- we need to fine-tune the boost values, currently  precise charms are usually before trusty. We still need to put trusty charms on the top, while "ordering" by download numbers within the same series.

File: internal/elasticsearch/query.go [hunk 1]
```diff
@@ -130,6 +130,26 @@ type BoostFactorFunction struct {
 	BoostFactor float64 `json:"boost_factor"`
 }
 
+// FieldValueFactorFunction boosts the results by the value of a field in the document.
+type FieldValueFactorFunction struct {
+	Field    string
+	Factor   float64
```
Comment (rogpeppe) [inline internal/elasticsearch/query.go]: If you added an "omitempty" tag to this field and the one below it, I think you could avoid the fv map in the MarshalJSON method below.

e.g. 

```
func (f FieldValueFactorFunction) MarshalJSON() ([]byte, error) {
    type fvff FieldValueFactorFunction
    return marshalNamedObject("field_value_factor", fvff(f))
}
```

**Comment by rogpeppe** [comment]
LGTM modulo Uros's QA requirements and one minor suggestion.

File: internal/elasticsearch/query.go (hunk 1 above)
Comment (mhilton) [inline internal/elasticsearch/query.go]: OK I'll give it a try.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/639/

File: internal/elasticsearch/query.go (hunk 1 above)
Comment (mhilton) [inline internal/elasticsearch/query.go]: done.

**Comment by mhilton** [comment]
In this new version I've dialed down the impact of download counts by a factor or 100.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/640/

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/641/

**Comment by urosj** [comment]
:+1: QA looks good. Please add that comment.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/642/

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/643/

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**